pysipp/launch.py

```diff
@@ -21,116 +22,109 @@ class TimeoutError(Exception):
-            proc = sp.Popen(
+            # proc = await trio.open_process(
+            proc = trio.Process(
                 shlex.split(cmd),
-                stdout=DEVNULL,
-                stderr=sp.PIPE
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.PIPE
             )
-            fd = proc.stderr.fileno()
-            log.debug("registering fd '{}' for pid '{}'".format(
-                fd, proc.pid))
-            fds2procs[fd] = self._procs[cmd] = proc
-            # register for stderr hangup events
-            self.poller.register(proc.stderr.fileno(), select.EPOLLHUP)
+            self._procs[cmd] = proc
+
             # limit launch rate
             time.sleep(1. / rate)
 
-        # launch waiter
-        self._waiter = threading.Thread(target=self._wait, args=(fds2procs,))
-        self._waiter.daemon = True
-        self._waiter.start()
-
-        return self.get(**kwargs) if block else self._procs
+        return self._procs
 
-    def _wait(self, fds2procs):
-        log.debug("started waiter for procs {}".format(fds2procs))
+    async def get(self, timeout=180):
+        '''Block up to `timeout` seconds for all agents to complete.
+        Either return (cmd, proc) pairs or raise `TimeoutError` on timeout
+        '''
         signalled = None
-        left = len(fds2procs)
-        collected = 0
-        while collected < left:
-            pairs = self.poller.poll()  # wait on hangup events
-            log.debug("received hangup for pairs '{}'".format(pairs))
-            for fd, status in pairs:
-                collected += 1
-                proc = fds2procs[fd]
-                # attach streams so they can be read more then once
-                log.debug("collecting streams for {}".format(proc))
-                proc.streams = Streams(*proc.communicate())  # timeout=2))
-                if proc.returncode != 0 and not signalled:
+
+        # taken mostly verbatim from ``trio.run_process()``
+        async def read_output(stream):
+            chunks = []
+            async with stream:
+                try:
+                    while True:
+                        chunk = await stream.receive_some(32768)
+                        if not chunk:
+                            break
+                        chunks.append(chunk)
+                except trio.ClosedResourceError:
+                    pass
+
+                return b"".join(chunks)
+
+        async def wait_on_proc(proc):
+            nonlocal signalled
+            async with proc as proc:
+                rc = await proc.wait()
+                if rc != 0 and not signalled:
                     # stop all other agents if there is a failure
                     signalled = self.stop()
 
-        log.debug("terminating waiter thread")
+                # collect stderr output
+                proc.stderr_output = await read_output(proc.stderr)
+
+        try:
+            with trio.fail_after(timeout):
```

Comment:
@kontsaki yeah i'm thinking we should maybe factor this out just so it's clear that it's a global timeout?